Adds cross platform architecture compilation for native image

kgston opened the pull request:
Closes #1443

This attempts to fill in the gap in cross platform architecture native image packaging. Added a new key `graalVMNativeImagePlatformArch: Option[String]` to specify the target platform architecture. Currently it only supports a single platform as doing multi platform would involve a much bigger change in the code base.

While this works technically, the performance is still rather lacking. When I crossed compiled (amd64 -> arm64) compile times took ~19x longer. The simple hello world test took over 5 minutes to produce a binary. 

The 5+ minutes test case also triggered a bug in [SBT #6771](https://github.com/sbt/sbt/issues/6671) causing the test to be aborted halfway through, but was recently fixed in SBT 1.9.3 hence the upgrade. Also updated the tests to use the `--no-fallback` option as I noticed that it was generating a fallback image when running the tests.

Also, it was a little tricky to correctly detect if an existing container was in the correct platform architecture when `graalVMNativeImagePlatformArch` is not defined. I used a docker image label to get around that since docker doesn't provide a way to report the host platform (using `uname -m` isn't ideal since it doesn't match docker's convention). The best it does is provide a special `local` value to signal building the image in the same platform architecture as the host.

lightbend-cla-validator commented:
Hi @kgston,

Thank you for your contribution! We really value the time you've taken to put this together.

We see that you have signed the Lightbend Contributors License Agreement before, however, the CLA has changed since you last signed it.
Please review the new CLA and sign it before we proceed with reviewing this pull request:

[https://www.lightbend.com/contribute/cla](https://www.lightbend.com/contribute/cla)

kgston commented:
The CI checks seems to be unhappy that I added a new parameter for specifying the target platform architecture.

guizmaii commented on .github/workflows/validate-pr.yml:
```diff
@@ -254,7 +254,7 @@ jobs:
     - name: Setup GraalVM environment
       uses: olafurpg/setup-scala@v10
```
This action is deprecated. I think it'd be better to migrate to https://github.com/graalvm/setup-graalvm

kgston commented:
@muuki88 Could you take a look at the PR when you find the time? Thanks.

muuki88 commented on src/main/scala/com/typesafe/sbt/packager/graalvmnativeimage/GraalVMNativeImagePlugin.scala:
```diff
@@ -166,35 +186,62 @@ object GraalVMNativeImagePlugin extends AutoPlugin {
           ExecCmd("RUN", "gu", "install", "native-image"),
           ExecCmd("RUN", "sh", "-c", "ln -s /opt/graalvm-ce-*/bin/native-image /usr/local/bin/native-image"),
-          ExecCmd("ENTRYPOINT", "native-image")
+          ExecCmd("CMD", "native-image")
```
I will never get the difference in my head. Why is this change from entrypoint to cmd needed?

muuki88 reviewed (approved):
Thanks for this high quality contribution 🙏

While I can't really give feedback on the graalvm implementation, the changes and tests look very good.

Only the entrypoint vs cmd is what I don't get 🫣 I'll merge if you give a short explanation here that I can reference if someone else wants to change it back 😁

muuki88 commented:
> The CI checks seems to be unhappy that I added a new parameter for specifying the target platform architecture.

IIRC there's another, private trait where those parameters can be added. If not, I'm fine with adding those binary compatibility breakages to the mima exceptions.

kgston commented:
> > The CI checks seems to be unhappy that I added a new parameter for specifying the target platform architecture.
> 
> IIRC there's another, private trait where those parameters can be added. If not, I'm fine with adding those binary compatibility breakages to the mima exceptions.

Hmm, I don't know if its exactly breaking binary compatibility since the old parameters still work the same as the old versions. According to the CI, it just says that the new param that I added doesn't exist on the older version, which is to be expected. 

Is there a better, more appropriate way to specify the target platform?

muuki88 commented:
I'm referencing this validation error: https://github.com/sbt/sbt-native-packager/actions/runs/5718905995/job/15495686495#step:7:1

which indicates binary compatibility issues. What are you referencing to?

kgston commented:
I've fixed the binary compat issue and all the subsequent issues that popped up from the graalvm test. Other unrelated tests are failing, and I assume it is a side effect from upgrading SBT. I'll try to debug the issue when I find some time, but taking a quick look at the github action logs, it is not immediately obvious what's the issue. Any hints would be appreciated, otherwise I might have to create a separate PR for upgrading to SBT 1.9.3 to try and isolate the issue.

muuki88 commented:
triggered a rerun .. the appveyor thing seems super weird. 

Do the tests fail on your local machine?

kgston commented:
I tried running `validateDebian`, `validateDocker` and `validateWindows` on `master` branch and I do get the same errors as with on the Github Actions too.. So it might be unrelated to the SBT 1.9.3 upgrade after all? 🤷‍♂️
